### PR TITLE
Fix bug comparing max_steps to global step which inits at 0

### LINKED ISSUE #4193: max_steps has no effect in combination with gradient accumulation
## 🐛 Bug

```python 

import os
import torch
from torch.utils.data import Dataset
from pytorch_lightning import Trainer, LightningModule


class RandomDataset(Dataset):
    def __init__(self, size, length):
        self.len = length
        self.data = torch.randn(length, size)

    def __getitem__(self, index):
        return self.data[index]

    def __len__(self):
        return self.len


class BoringModel(LightningModule):

    def __init__(self):
        """
        Testing PL Module

        Use as follows:
        - subclass
        - modify the behavior for what you want

        class TestModel(BaseTestModel):
            def training_step(...):
                # do your own thing

        or:

        model = BaseTestModel()
        model.training_epoch_end = None

        """
        super().__init__()
        self.layer = torch.nn.Linear(32, 2)

    def forward(self, x):
        return self.layer(x)

    def loss(self, batch, prediction):
        # An arbitrary loss to have a loss that updates the model weights during `Trainer.fit` calls
        return torch.nn.functional.mse_loss(prediction, torch.ones_like(prediction))

    def step(self, x):
        x = self.layer(x)
        out = torch.nn.functional.mse_loss(x, torch.ones_like(x))
        return out

    def training_step(self, batch, batch_idx):
        output = self.layer(batch)
        loss = self.loss(batch, output)
        return {"loss": loss}

    def training_step_end(self, training_step_outputs):
        return training_step_outputs

    def training_epoch_end(self, outputs) -> None:
        torch.stack([x["loss"] for x in outputs]).mean()

    def validation_step(self, batch, batch_idx):
        output = self.layer(batch)
        loss = self.loss(batch, output)
        return {"x": loss}

    def validation_epoch_end(self, outputs) -> None:
        torch.stack([x['x'] for x in outputs]).mean()

    def test_step(self, batch, batch_idx):
        output = self.layer(batch)
        loss = self.loss(batch, output)
        return {"y": loss}

    def test_epoch_end(self, outputs) -> None:
        torch.stack([x["y"] for x in outputs]).mean()

    def configure_optimizers(self):
        optimizer = torch.optim.SGD(self.layer.parameters(), lr=0.1)
        lr_scheduler = torch.optim.lr_scheduler.StepLR(optimizer, step_size=1)
        return [optimizer], [lr_scheduler]


def run_test():
    class TestModel(BoringModel):

        def on_train_epoch_start(self) -> None:
            print('override any method to prove your bug')

    # fake data
    train_data = torch.utils.data.DataLoader(RandomDataset(32, 64))
    val_data = torch.utils.data.DataLoader(RandomDataset(32, 64))
    test_data = torch.utils.data.DataLoader(RandomDataset(32, 64))

    # model
    model = TestModel()
    trainer = Trainer(
        default_root_dir=os.getcwd(),
        # -----------------------------------------------------------------------------
        max_steps=10,  # HERE HAS NO EFFECT IN COMBINATION WITH accumulate_grad_batches
        # -----------------------------------------------------------------------------
        accumulate_grad_batches=2,
        weights_summary=None,
    )
    trainer.fit(model, train_data, val_data)
    trainer.test(test_dataloaders=test_data)


if __name__ == '__main__':
    run_test()


```

This training runs for all 1000 epochs instead of just max steps.

### Expected behavior

Training runs for max_steps. 

### Additional context

The tests with TODO marker can be found here  #4190

### PR DESCRIPTION
**Pull request by SeanNaren**
## What does this PR do?

Fixes #4193 

Just going to make a test before calling this ready for review.

## PR review
 - [ ] Is this pull request ready for review? (if not, please submit in draft mode)

Anyone in the community is free to review the PR once the tests have passed.    
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Review by justusschock — APPROVED** [review]

File: pytorch_lightning/trainer/trainer.py [hunk 1]
```diff
@@ -481,7 +481,7 @@ def train(self):
                 # run train epoch
                 self.train_loop.run_training_epoch()
 
-                if self.max_steps and self.max_steps <= self.global_step:
+                if self.max_steps and self.max_steps <= self.global_step + 1:
```
Comment (awaelchli) [inline pytorch_lightning/trainer/trainer.py:484]: this can't be the fix, since this check here is at the end of each epoch. 
The fix must certainly be in the loop that runs over the training batches.

**Review by awaelchli — CHANGES_REQUESTED** [review]

File: pytorch_lightning/trainer/trainer.py (hunk 1 above)
Comment (SeanNaren) [reply pytorch_lightning/trainer/trainer.py:484]: once the training epoch returns after reaching max steps, we need to ensure we break out of the loop. this is what the fix does. you can see the logic is incorrect as it doesn't reflect the logic here:

https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/trainer/training_loop.py#L582

Comment (SeanNaren) [reply pytorch_lightning/trainer/trainer.py:484]: To summarise, the current PR fixes the issue by handling the assumption that with accumulate_grad_batches of 1 this line will always increment the global step: https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/trainer/training_loop.py#L619

Now the check will prevent the step from going into an infinite loop as the global step is never incremented (when accumulate_grad_batches is greater than 1. 

When accumulate_grad_batches is greater than 1, this causes an infinite loop as the global step is never incremented and the checks are not in sync.

@awaelchli I'll double check the other test you also linked!

Comment (awaelchli) [reply pytorch_lightning/trainer/trainer.py:484]: ok, let's see. 
I merged master and uncommented my tests. Let's see if they work as intended :)

Comment (SeanNaren) [reply pytorch_lightning/trainer/trainer.py:484]: Seems you are right, we're not in an infinite loop but we're missing a step. Will investigate further

File: tests/trainer/optimization/test_backward_calls.py [hunk 2]
```diff
@@ -48,8 +47,6 @@ def test_backward_count_with_closure(torch_backward):
 
     torch_backward.reset_mock()
 
-    # TODO: max_steps seems to fail together with accumulation
-    # trainer = Trainer(max_steps=5, accumulate_grad_batches=2
-    trainer = Trainer(max_epochs=1, limit_train_batches=5, accumulate_grad_batches=2)
+    trainer = Trainer(max_steps=5, accumulate_grad_batches=2)
     trainer.fit(model)
-    assert torch_backward.call_count == 5
+    assert torch_backward.call_count == 10
```
Comment (awaelchli) [inline tests/trainer/optimization/test_backward_calls.py:52]: I agree this is correct. My counting was not correct when I wrote the test, saw it yesterday but forgot to tell you. Thanks for updating!

File: tests/trainer/test_trainer.py [hunk 3]
```diff
@@ -661,6 +661,26 @@ def test_trainer_min_steps_and_epochs(tmpdir):
...
 
 
+def test_trainer_max_steps_accumulate_batches(tmpdir):
+    """Verify model trains according to specified max steps with grad accumulated batches"""
+    model, trainer_options, num_train_samples = _init_steps_model()
+
+    # define less train steps than epochs
+    trainer_options.update(
+        default_root_dir=tmpdir,
+        max_steps=num_train_samples + 10,
```
Comment (awaelchli) [inline tests/trainer/test_trainer.py:671]:
```suggestion
        max_steps=(num_train_samples + 10),
```
for visual clarity

File: pytorch_lightning/trainer/training_loop.py [hunk 4]
```diff
@@ -834,13 +837,19 @@ def run_on_epoch_end_hook(self, epoch_output):
...
         # progress global step according to grads progress
         if num_accumulated_batches_reached or num_training_batches_reached:
             self.trainer.global_step += 1
 
+    def accumulated_batches_reached(self):
+        return (self.trainer.batch_idx + 1) % self.trainer.accumulate_grad_batches == 0
+
+    def num_training_batches_reached(self):
+        return (self.trainer.batch_idx + 1) == self.trainer.num_training_batches
+
```
Comment (awaelchli) [inline pytorch_lightning/trainer/training_loop.py:852]: nice <3

**Review by awaelchli — APPROVED** [review]

**Review by tchaton — APPROVED** [review]
Great PR ! Nice cleaned up.

File: pytorch_lightning/trainer/training_loop.py [hunk 5]
```diff
@@ -834,13 +837,19 @@ def run_on_epoch_end_hook(self, epoch_output):
...
-        num_accumulated_batches_reached = (self.trainer.batch_idx + 1) % self.trainer.accumulate_grad_batches == 0
-        num_training_batches_reached = (self.trainer.batch_idx + 1) == self.trainer.num_training_batches
+        num_accumulated_batches_reached = self.accumulated_batches_reached()
+        num_training_batches_reached = self.num_training_batches_reached()
 
         # progress global step according to grads progress
         if num_accumulated_batches_reached or num_training_batches_reached:
             self.trainer.global_step += 1
 
+    def accumulated_batches_reached(self):
```
Comment (Borda) [inline pytorch_lightning/trainer/training_loop.py:847]: can we have it rather as protected as it is not meant to be used by the user...

Comment (SeanNaren) [reply pytorch_lightning/trainer/training_loop.py:847]: Will make a followup PR
